numpy/core/src/multiarray/hashdescr.c

```diff
@@ -224,7 +230,12 @@ static int _array_descr_walk(PyArray_Descr* descr, PyObject *l)
                         "(Hash) fields is not a dict ???");
                 return -1;
             }
-            st = _array_descr_walk_fields(descr->fields, l);
+            if (!PyTuple_Check(descr->names)) {
```

Comment:
Oh, so here is the check I was missing before... I think it will be clearer if we moved this one into `_array_descr_walk_fields`. Perhaps we can also move the check for `fields` into it?